rust/src/shards/gui/widgets/text_input.rs

```diff
@@ -0,0 +1,309 @@
+      let var = Var::ephemeral_string(str.as_str());
+      cloneVar(self, &var);
+    }
+
+    text.chars().count()
+  }
+
+  fn delete_char_range(&mut self, char_range: std::ops::Range<usize>) {
+    assert!(char_range.start <= char_range.end);
+
+    let byte_start = self.byte_index_from_char_index(char_range.start);
+    let byte_end = self.byte_index_from_char_index(char_range.end);
+
+    if byte_start == byte_end {
+      // nothing to do
+      return;
+    }
+
+    unsafe {
+      let current_len = self.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen as usize;
+      let base_ptr =
+        self.payload.__bindgen_anon_1.__bindgen_anon_2.stringValue as *mut std::os::raw::c_char;
+      // move rest of the text at the deletion location
+      std::ptr::copy(
+        base_ptr.add(byte_end),
+        base_ptr.add(byte_start),
+        current_len - byte_end,
+      );
+      // update the length
+      let new_len = current_len - byte_end + byte_start;
+      self.payload.__bindgen_anon_1.__bindgen_anon_2.stringLen = new_len as u32;
+      // fixup null-terminator
+      *base_ptr.add(new_len) = 0;
+    }
+  }
+}
+
+impl Shard for TextInput {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.TextInput")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.TextInput-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.TextInput"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &NONE_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &STRING_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&TEXTINPUT_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.variable.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.variable.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if self.variable.is_variable() {
+      self.should_expose = true; // assume we expose a new variable
+
+      let shared: ExposedTypes = data.shared.into();
+      for var in shared {
+        let (a, b) = unsafe {
+          (
+            CStr::from_ptr(var.name),
+            CStr::from_ptr(self.variable.get_name()),
+          )
+        };
+        if CStr::cmp(&a, &b) == Ordering::Equal {
+          self.should_expose = false;
+          self.mutable_text = var.isMutable;
+          if var.exposedType.basicType != shardsc::SHType_String {
+            return Err("TextInput: string variable required.");
+          }
+          break;
+        }
+      }
+    } else {
+      self.mutable_text = false;
+    }
+
+    Ok(common_type::string)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
```

Comment:
Ah so you can write `(TextInput .variable)` without having to assign it before?

Reply:
Yes it's a supported scenario.